### ISSUE TITLE
CSV Logger acts weirdly in Callbacks

### ISSUE THREAD

**Issue by oabuhamdan** [Lightning-AI/pytorch-lightning#20110]
### Bug description

I use CSV logger inside callbacks as the following
```
pl_module.log('epoch_throughput', throughput, on_epoch=True, logger=True, sync_dist=True, reduce_fx="sum")
pl_module.log('epoch_time', epoch_time, logger=True, on_epoch=True, sync_dist=True)
```
When I use two callbacks to log data at the same time point (i.e. on_train_epoch_end), it logs the data into two rows
![image](https://github.com/user-attachments/assets/cc05654c-af1f-490b-8c74-104ca1aed632)

### What version are you seeing the problem on?

master

### How to reproduce the bug

```python
class CallbackX(Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        self.epoch_start_time = time.time()

    def on_train_epoch_end(self, trainer, pl_module):
        epoch_time = time.time() - self.epoch_start_time
        pl_module.log('epoch_time', epoch_time, logger=True, on_epoch=True, sync_dist=True)


class CallbackY(Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        self.epoch_start_time = time.time()

    def on_train_epoch_end(self, trainer, pl_module):
        epoch_time = time.time() - self.epoch_start_time
        pl_module.log('epoch_time2', epoch_time, logger=True, on_epoch=True, sync_dist=True)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
	- GPU:               Tesla T4
	- available:         False
	- version:           12.1
* Lightning:
	- lightning:         2.3.0
	- lightning-utilities: 0.11.2
	- pytorch-lightning: 2.3.0
	- torch:             2.3.1
	- torch-tb-profiler: 0.4.3
	- torchmetrics:      1.4.0.post0
	- torchvision:       0.18.1
* Packages:
	- absl-py:           2.1.0
	- aiohttp:           3.9.5
	- aiosignal:         1.3.1
	- annotated-types:   0.7.0
	- attrs:             23.2.0
	- certifi:           2024.6.2
	- cffi:              1.16.0
	- charset-normalizer: 3.3.2
	- cloudpickle:       3.0.0
	- deepspeed:         0.14.4
	- dool:              1.3.2
	- filelock:          3.15.4
	- frozenlist:        1.4.1
	- fsspec:            2024.6.0
	- graphviz:          0.8.4
	- grpcio:            1.64.1
	- gviz-api:          1.10.0
	- hjson:             3.1.0
	- idna:              3.7
	- jinja2:            3.1.4
	- lightning:         2.3.0
	- lightning-utilities: 0.11.2
	- markdown:          3.6
	- markupsafe:        2.1.5
	- mpmath:            1.3.0
	- multidict:         6.0.5
	- mxnet:             1.9.1
	- networkx:          3.3
	- ninja:             1.11.1.1
	- numpy:             1.25.0
	- nvidia-cublas-cu12: 12.1.3.1
	- nvidia-cuda-cupti-cu12: 12.1.105
	- nvidia-cuda-nvrtc-cu12: 12.1.105
	- nvidia-cuda-runtime-cu12: 12.1.105
	- nvidia-cudnn-cu12: 8.9.2.26
	- nvidia-cufft-cu12: 11.0.2.54
	- nvidia-curand-cu12: 10.3.2.106
	- nvidia-cusolver-cu12: 11.4.5.107
	- nvidia-cusparse-cu12: 12.1.0.106
	- nvidia-ml-py:      12.555.43
	- nvidia-nccl-cu12:  2.20.5
	- nvidia-nvjitlink-cu12: 12.5.40
	- nvidia-nvtx-cu12:  12.1.105
	- packaging:         24.1
	- pandas:            2.2.2
	- pillow:            10.3.0
	- pip:               24.1
	- protobuf:          4.25.3
	- psutil:            6.0.0
	- py-cpuinfo:        9.0.0
	- pycparser:         2.22
	- pydantic:          2.7.4
	- pydantic-core:     2.18.4
	- python-dateutil:   2.9.0.post0
	- pytorch-lightning: 2.3.0
	- pytz:              2024.1
	- pyyaml:            6.0.1
	- requests:          2.32.3
	- setuptools:        65.5.0
	- six:               1.16.0
	- sympy:             1.12.1
	- tensorboard:       2.17.0
	- tensorboard-data-server: 0.7.2
	- tensorboard-plugin-profile: 2.15.1
	- tensorboardx:      2.6.2.2
	- torch:             2.3.1
	- torch-tb-profiler: 0.4.3
	- torchmetrics:      1.4.0.post0
	- torchvision:       0.18.1
	- tqdm:              4.66.4
	- triton:            2.3.1
	- typing-extensions: 4.12.2
	- tzdata:            2024.1
	- urllib3:           2.2.2
	- werkzeug:          3.0.3
	- wheel:             0.43.0
	- yarl:              1.9.4
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.11.7
	- release:           3.10.0-1127.19.1.el7.x86_64
	- version:           #1 SMP Tue Aug 25 17:23:54 UTC 2020

</details>

**Comment by SkafteNicki** [comment]
Hi, thanks for raising this issue. Not sure in what version this was fixed, but after trying to reproduce the issue, I can report that it seems fixed. Values logged in different callbacks are logged to the same line in csv. Script for reproduction:

```python
import time
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, TensorDataset
import pytorch_lightning as pl
from pytorch_lightning.callbacks import Callback
from pytorch_lightning.loggers import CSVLogger


class CallbackX(Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        self.epoch_start_time = time.time()

    def on_train_epoch_end(self, trainer, pl_module):
        epoch_time = time.time() - self.epoch_start_time
        pl_module.log('epoch_time', epoch_time, logger=True, on_epoch=True, sync_dist=True)


class CallbackY(Callback):
    def on_train_epoch_start(self, trainer, pl_module):
        self.epoch_start_time = time.time()

    def on_train_epoch_end(self, trainer, pl_module):
        epoch_time = time.time() - self.epoch_start_time
        pl_module.log('epoch_time2', epoch_time, logger=True, on_epoch=True, sync_dist=True)


class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = nn.Linear(10, 1)
        
    def forward(self, x):
        return self.layer(x)
    
    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        self.log('train_loss', loss)
        return loss
    
    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.01)


def main():
    # Create dummy data
    x = torch.randn(100, 10)
    y = torch.randn(100, 1)
    dataset = TensorDataset(x, y)
    train_dataloader = DataLoader(dataset, batch_size=32)
    
    # Create model
    model = SimpleModel()
    
    # Create CSV logger
    csv_logger = CSVLogger("logs", name="issue_20110_simple")
    
    # Create two callbacks that both log at on_train_epoch_end
    callback_x = CallbackX()
    callback_y = CallbackY()
    
    # Create trainer
    trainer = pl.Trainer(
        max_epochs=2,
        callbacks=[callback_x, callback_y],
        logger=csv_logger,
        accelerator='cpu',
        devices=1,
        enable_progress_bar=False
    )
    
    # Train
    trainer.fit(model, train_dataloader)
    

if __name__ == "__main__":
    main()
```
Closing issue, but feel free to ping and reopen if necessary.
